src/java/org/apache/cassandra/security/SSLFactory.java

```diff
@@ -397,41 +392,121 @@ public static synchronized void initHotReloading(EncryptionOptions.ServerEncrypt
+                                           Set<String> supportedCiphers)
         {
-            // Ensure we're able to create both server & client SslContexts if they might ever be needed
-            if (clientOpts != null && clientOpts.tlsEncryptionPolicy() != EncryptionOptions.TlsEncryptionPolicy.UNENCRYPTED)
+            Objects.requireNonNull(defaultCiphers, "defaultCiphers");
+            Objects.requireNonNull(supportedCiphers, "supportedCiphers");
+
+            final List<String> newCiphers;
+            if (ciphers == null)
+            {
+                newCiphers = new ArrayList<>(defaultCiphers.size());
+                ciphers = defaultCiphers;
+            }
+            else
+            {
+                newCiphers = new ArrayList<>(supportedCiphers.size());
+            }
+            for (String c : ciphers)
             {
-                createNettySslContext(clientOpts, clientOpts.require_client_auth, SocketType.SERVER, openSslIsAvailable());
-                createNettySslContext(clientOpts, clientOpts.require_client_auth, SocketType.CLIENT, openSslIsAvailable());
+                if (c == null)
+                {
+                    break;
+                }
+                if (supportedCiphers.contains(c))
+                {
+                    newCiphers.add(c);
+                }
+                else
+                {
+                    if (settingDescription != null)
+                    {
+                        logger.warn("Dropping unsupported cipher_suite {} from {} configuration",
+                                    c, settingDescription.toLowerCase());
+                    }
+                }
             }
+            if (newCiphers.isEmpty())
+            {
+                throw new IllegalStateException("No ciphers left after filtering supported cipher suite");
+            }
+
+            return newCiphers.toArray(new String[0]);
         }
-        catch (Exception e)
+    }
+
+    public static void validateSslContext(String contextDescription, EncryptionOptions options, boolean buildTrustStore, boolean logProtocolAndCiphers) throws IOException
+    {
+        if (options != null && options.tlsEncryptionPolicy() != EncryptionOptions.TlsEncryptionPolicy.UNENCRYPTED)
         {
-            throw new IOException("Failed to create SSL context using client_encryption_options!", e);
+            try
+            {
+                CipherSuiteFilter loggingCipherSuiteFilter = logProtocolAndCiphers ? new LoggingCipherSuiteFilter(contextDescription)
+                                                                                   : LoggingCipherSuiteFilter.QUIET_FILTER;
+                SslContext sslContext = createNettySslContext(options, buildTrustStore, SocketType.SERVER, openSslIsAvailable(), loggingCipherSuiteFilter);
+                SSLEngine engine = sslContext.newEngine(ByteBufAllocator.DEFAULT);
```

Comment:
don't we need to release as you don't actually use this? its ref counted and `io.netty.handler.ssl.ReferenceCountedOpenSslEngine` looks like it should be cleaned up if allocated.

Reply:
Thanks for spotting this. Pushed up the fix we discussed.